File: Design-Patterns-Using-Java/Prototype/PrototypePattern.java
import java.util.HashMap;
import java.util.Map;

abstract class Prototype implements Cloneable {
    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    abstract void display();
}

class ConcretePrototype1 extends Prototype {
    @Override
    void display() {
        System.out.println("ConcretePrototype1");
    }
}

class ConcretePrototype2 extends Prototype {
    @Override
    void display() {
        System.out.println("ConcretePrototype2");
    }
}

class PrototypeRegistry {
    private static final Map<String, Prototype> prototypes = new HashMap<>();

    static void addPrototype(String key, Prototype value) {
        prototypes.put(key, value);
    }

    static Prototype getPrototype(String key) throws CloneNotSupportedException {
        if (prototypes.containsKey(key)) {
            return (Prototype) prototypes.get(key).clone();
        }
        return null;
    }

    static void load() {
        addPrototype("CP1", new ConcretePrototype1());
        addPrototype("CP2", new ConcretePrototype2());
    }
}

public class PrototypePattern {
    public static void main(String[] args) throws CloneNotSupportedException {
        PrototypeRegistry.load();
        Prototype c1 = PrototypeRegistry.getPrototype("CP1");
        Prototype c2 = PrototypeRegistry.getPrototype("CP2");
        c1.display();
        c2.display();
    }
}

/*
ConcretePrototype1
ConcretePrototype2
*/

File: Design-Patterns-Using-Kotlin/Prototype/PrototypePattern.kt
abstract class Prototype : Cloneable {
    @Throws(CloneNotSupportedException::class)
    public override fun clone(): Any {
        return super.clone()
    }

    abstract fun display()
}

class ConcretePrototype1 : Prototype() {
    override fun display() {
        println("ConcretePrototype1")
    }
}

class ConcretePrototype2 : Prototype() {
    override fun display() {
        println("ConcretePrototype2")
    }
}

object PrototypeRegistry {
    private val prototypes: MutableMap<String, Prototype> = HashMap()

    init {
        load()
    }

    fun addPrototype(key: String, value: Prototype) {
        prototypes[key] = value
    }

    @Throws(CloneNotSupportedException::class)
    fun getPrototype(key: String): Prototype? {
        if (prototypes.containsKey(key)) {
            return prototypes[key]!!.clone() as Prototype
        }
        return null
    }

    fun load() {
        addPrototype("CP1", ConcretePrototype1())
        addPrototype("CP2", ConcretePrototype2())
    }
}

// Client Code
fun main() {
    PrototypeRegistry.load()
    val c1 = PrototypeRegistry.getPrototype("CP1")
    val c2 = PrototypeRegistry.getPrototype("CP2")
    c1!!.display()
    c2!!.display()
}

/*
ConcretePrototype1
ConcretePrototype2
*/

